src/v/s3/client.cc

```diff
@@ -606,9 +606,10 @@ ss::future<> client::put_object(
     auto header = _requestor.make_unsigned_put_object_request(
       name, id, payload_size, tags);
     if (!header) {
-        return body.close().then([&header] {
-            return ss::make_exception_future<>(
-              std::system_error(header.error()));
+        return ss::do_with(header.error(), [&body](auto& error) {
+            return body.close().then([&error] {
+                return ss::make_exception_future<>(std::system_error(error));
+            });
```

Comment:
this looks like it fixes the life time issue, but it looks like you could make it even more light weight. would the following also work?

```diff
diff --git a/src/v/s3/client.cc b/src/v/s3/client.cc
index b7a2af01a..af9f1a146 100644
--- a/src/v/s3/client.cc
+++ b/src/v/s3/client.cc
@@ -606,7 +606,7 @@ ss::future<> client::put_object(
     auto header = _requestor.make_unsigned_put_object_request(
       name, id, payload_size, tags);
     if (!header) {
-        return body.close().then([&header] {
+        return body.close().then([header] {
             return ss::make_exception_future<>(
               std::system_error(header.error()));
         });
```